### PR TITLE
ecds: delay readiness until the worker update is complete

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: Delays ECDS init readiness until the worker update is complete.

Additional Description: ECDS triggers `ready()` as soon as xDS update is received. This in turn activates the parent listener, which causes a worker to accept requests before the filter config is published to that worker. This is especially problematic for Wasm filters, which can take up to seconds to fully initialize. The solution is to move the ready call after the filter config is created and published to the workers. The worker queues guarantee that the new config is available before the warmed parent (listener) resource is published in the same queues.

Risk Level: medium, change in lifecycle
Testing: manual, unit
Docs Changes: none
Release Notes: yes
Fixes: https://github.com/envoyproxy/envoy/issues/25946

### CODE REVIEW

File: source/common/common/cleanup.h
Comment (kyessenov) [inline source/common/common/cleanup.h:15]: cc @ravenblackx

**Comment by kyessenov** [comment]
@KBaichoo do you want to be a first pass reviewer?

**Review by adisuissa — COMMENTED** [review]
Thanks for solving this problem!
This is the first time I see that there is an expectation to wait until all worker threads are updated after an xDS update. A few high-level questions:
1. This only impacts ECDS. Should the same fix be applied to normal filters (say an LDS response that has a WASM filter)?
2. It isn't clear to me that they waiting time to update all the worker threads is bounded. Could there be a starvation here? Why is this different than failing to receive an ECDS response, or receiving one with an invalid config?
3. This will work also for non-WASM filters. Is there an expectation on all filters initialization time?
4. There are other cases where one needs to wait until Envoy is "ready" before diverting traffic to it. Can the `config_reload` counter be used to ensure that the worker threads applied the new config?

**Comment by kyessenov** [comment]
> 1. This only impacts ECDS. Should the same fix be applied to normal filters (say an LDS response that has a WASM filter)?
I think this is a problem with xDS dependencies, and Wasm in LDS doesn't have this issue because the entire listener is fully constructed when it's applied. The main thread is blocked while Wasm initializes so it's completely set-up when posted to workers for accepting requests.

> 2. It isn't clear to me that they waiting time to update all the worker threads is bounded. Could there be a starvation here? Why is this different than failing to receive an ECDS response, or receiving one with an invalid config?

There's a risk of starvation. I'm not sure this is any different from holding LDS warm until RDS/CDS arrives. The issue is that we mark ready on accepting the config, not applying the config. Wasm just takes a really long time to apply. It'd be good to make sure Wasm does heavy computation on a different thread than main if it doesn't do so already.

> 3. This will work also for non-WASM filters. Is there an expectation on all filters initialization time?

This is a problem with ECDS, not Wasm. The customer was using a Wasm filter for telemetry, and getting 4xx or 5xx for an "optional" filter is not right, when explicitly requesting to "warm" the filter.

> 4. There are other cases where one needs to wait until Envoy is "ready" before diverting traffic to it. Can the `config_reload` counter be used to ensure that the worker threads applied the new config?

Yeah, that's why it was added. We missed the "warming" issue when we added the `config_reload` counter.

File: source/common/filter/config_discovery_impl.cc
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.cc:131]: You mean `onConfigUpdate` that throws will be followed by `onConfigFailed`? I thought `onConfigFailed` was about the failure to receive any config, not receive an invalid config.

File: source/server/server.h
Comment (kyessenov) [inline source/server/server.h:373]: dependency on lifecycle notifier requires changing the order of destruction.

**Comment by alyssawilk** [comment]
Looks like this needs a main merge, but @adisuissa should it be tagged as waiting or is it waiting on your review?

**Review by adisuissa — COMMENTED** [review]
My main concern (which may be wrong) is starvation.
In xDS-warming, there's a timer for receiving a the dependent resource, and if the timer is triggered, the dependent resource is marked "removed" and the parent is notified.
In this case, IIUC, the timer will be removed (as an xDS update was received), but now the filter initialization step can take an unbounded amount of time. This could lead to unresponsive Envoys during bootstrap.
This could lead to having a problematic filter config that has wide impact on the entire Envoy behavior.

Will it be possible for all worker threads to signal when they are ready, and for the Envoy bootstrap to wait for it (or for a timeout)? This can be done with an specific bootstrap config. WDYT?

File: source/common/filter/config_discovery_impl.cc
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.cc:131]: Thanks for letting me know, this should simplify the code.

**Comment by kyessenov** [comment]
/retest

@adisuissa Did I answer your concern about starvation? I don't think initial startup is affected at all, we special case pre-worker sequence since there are no active workers then.

File: source/common/filter/config_discovery_impl.cc
Comment (adisuissa) [inline source/common/filter/config_discovery_impl.cc:131]: nit: consider std::move

File: source/common/filter/config_discovery_impl.h
Comment (adisuissa) [inline source/common/filter/config_discovery_impl.h:344]: nit: should this be ConfigVersionConstSharedPtr (I didn't see an update of the `last_` data members)

**Review by adisuissa — COMMENTED** [review]
Apologies for the late response.

IMHO it is not ideal that ECDS needs to know the state outside its context (whether worker threads have been initialized or not) and that it changes its mode of operation (i.e, signaling when it is ready). It would be better to have ECDS tell that it is ready (whenever that is), and the component it notifies that will know about the external state will take appropriate action.

If introducing the workers state into ECDS is the desired state, then these code changes LGTM.

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.h]: done

File: source/common/filter/config_discovery_impl.cc
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.cc]: done

**Comment by kyessenov** [comment]
@adisuissa Can I have another stamp? I reverted some test changes from previous attempts.

File: source/common/filter/config_discovery_impl.h
Comment (adisuissa) [inline source/common/filter/config_discovery_impl.h:345]: Upon reading the code again, I think this actually doesn't need to be a shared-ptr, but just a struct. It will be similar to the previous fields, but grouped in a struct. It will save a bit of memory, and will cause less confusion (avoiding the need to understand why is this a shared-ptr).

**Review by adisuissa — DISMISSED** [review]
LGTM!
Left a minor comment, that can be cleaned up in a future PR.

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.h:345]: I tried it with struct but the issue is that config is captured in the lambda for filter config providers. Capturing by value is not possible because of the `MessagePtr` inside it.

**Comment by kyessenov** [comment]
Thanks again, resolved changelog conflict, but need an approval again.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
